PurchasesCoreSwift/Public/PurchaserInfo.swift

```diff
@@ -0,0 +1,317 @@
+            }
+
+            guard let firstSeenDateString = subscriberData["first_seen"] as? String,
+                  let firstSeenDate = PurchaserInfo.parseDate(date: firstSeenDateString, withDateFormatter: DateFormatter.dateFormatter) else {
+                return nil
+            }
+            self.firstSeen = firstSeenDate
+
+            guard let originalAppUserIdString = subscriberData["original_app_user_id"] as? String else {
+                return nil
+            }
+            self.originalAppUserId = originalAppUserIdString
+
+            if let managementUrlString = subscriberData["management_url"] as? String {
+                self.managementURL = parseURL(url: managementUrlString)
+            } else {
+                self.managementURL = nil
+            }
+
+            // Purchases and entitlements
+            self.nonSubscriptions = subscriberData["non_subscriptions"] as? [String: [[String: Any]]] ?? [String: [[String: Any]]]()
+            self.entitlements = subscriberData["entitlements"] as? [String: Any] ?? [String: Any]()
+            self.nonSubscriptionTransactions = TransactionsFactory().nonSubscriptionTransactions(withSubscriptionsData: nonSubscriptions, dateFormatter: DateFormatter.dateFormatter)
+
+            let latestNonSubscriptionTransactionsByProductId = [String: Any](uniqueKeysWithValues: nonSubscriptions.map { productId, transactionsArray in
+                (productId, transactionsArray.last)
+            })
+
+            self.allPurchases = latestNonSubscriptionTransactionsByProductId.merging(subscriptions) { (current, _) in current }
+
+            self.expirationDatesByProduct = PurchaserInfo.parseExpirationDate(expirationDates: subscriptions)
+            self.purchaseDatesByProduct = PurchaserInfo.parsePurchaseDate(purchaseDates: allPurchases)
+        }
+    }
+
+}
+
+private extension PurchaserInfo {
+
+    func activeKeys(dates: [String: Date?]) -> Set<String> {
+        return Set<String>(dates.keys.filter({
+            if let nonNullDate = dates[$0] as? Date {
+                return isAfterReferenceDate(date: nonNullDate)
+            }
+            return true
+        }))
+    }
+
+    func isAfterReferenceDate(date: Date) -> Bool { date.timeIntervalSince(self.requestDate) > 0 }
+
+    class func parseURL(url: Any) -> URL? {
+        if let urlString = url as? String {
+            return URL.init(string: urlString)
+        }
+        return nil
+    }
+
+    class func parseDate(date: Any, withDateFormatter: DateFormatter) -> Date? {
+        if let dateString = date as? String {
+            return DateFormatter.dateFormatter.date(from: dateString)
+        }
+        return nil
+    }
+
+    class func parseExpirationDate(expirationDates: [String: Any]) -> [String: Date?] { parseDatesIn(dates: expirationDates, label: "expires_date") }
+
+    class func parsePurchaseDate(purchaseDates: [String: Any]) -> [String: Date?] { parseDatesIn(dates: purchaseDates, label: "purchase_date") }
```

Comment:
we should avoid doing single-liners if the lines go over 120 chars, could you add a line break?

Reply:
should lint have caught this? or is that defined somewhere in particular?

Reply:
I think we might have disabled the rule at first since we already had a decent amount of code before adding the linter